File: Reqnroll/Chapter21/21.3-TestFileSystemPattern/After/WIMP.Specs/Support/Hooks.cs
using Microsoft.Extensions.Logging;

using Reqnroll;

using WIMP.Specs.Support.Logging;

namespace WIMP.Specs.Support;

[Binding]
public class Hooks(AppHostingContext appHostingContext, AppLogContext appLogContext, ReqnrollLoggerProvider reqnrollLoggerProvider,
    ILoggerFactory loggerFactory, IScenarioContext scenarioContext, IReqnrollOutputHelper outputHelper,
    TestFileSystem testFileSystem)
{
    private ILogger Logger => loggerFactory.CreateLogger(GetType());

    [AfterScenario(Order = 0)]
    public void SaveAppLogOnError()
    {
        if (scenarioContext.ScenarioExecutionStatus == ScenarioExecutionStatus.TestError)
        {
            string outputPath = Path.Combine(testFileSystem.OutputFolder,
                testFileSystem.GetScenarioSpecificFileName(".log"));
            appLogContext.SaveToFile(outputPath);
            Logger.LogInformation("Saved app log to {AppLogFile}", outputPath);
            outputHelper.AddAttachment(outputPath);
        }
    }

    [BeforeScenario]
    public void CreateAppHost()
    {
        appHostingContext.AppHost = new WimpAppHost(appLogContext, reqnrollLoggerProvider);
    }

    [AfterScenario]
    public void DisposeAppHost()
    {
        appHostingContext.AppHost.Dispose();
    }
}

File: CucumberJVM/Chapter18/18.4-SeededDataPattern/Before/src/test/java/com/wimp/app/specs/support/Hooks.java
package com.wimp.app.specs.support;

import com.wimp.app.specs.drivers.DatabaseDriver;
import io.cucumber.java.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Hooks {
    private static final Logger log = LoggerFactory.getLogger(Hooks.class);

    private final DatabaseDriver databaseDriver;

    public Hooks(DatabaseDriver databaseDriver) {
        this.databaseDriver = databaseDriver;
    }

    @Before(order = 0)
    public void resetDatabase() {
        databaseDriver.emptyDatabase();
    }
}
